### DISCUSSION TITLE
Stance detection in tweets (in favor, neutral, or against topic)

### DISCUSSION THREAD

**Discussion by metalaureate** [explosion/spaCy#11692] [Help: Coding & Implementations] [answered]
I am trying to use SpaCy to accomplish stance detection in tweets but a bit stuck for practical previous example work to adapt to my use case, where for a given tweet or sequence of tweets I can infer whether the author is in favor, neutral, or against a recognized named entity. The problem is subtly very different from sentiment analysis. e.g.

"Ukrainian refugees are suffering" would score a negative sentiment for Ukrainian refugees which is the complete opposite of the favorable stance implied by the author of the tweet. "Pregnant women have rights" is neutral for sentiment analysis, but clearly a positive stance. (e.g. see https://huggingface.co/finiteautomata/bertweet-base-sentiment-analysis?text=Pregnant+women+have+rights. )

I am using NER Tweets tutorial for the NER part of the solution (https://github.com/explosion/projects/tree/576ba0089cdca320a9f56cde9f4220f12a1f6000/tutorials/ner_tweets)

but how to infer stance (favor, neutral, or against)

I found  an article SemEval-2016 Task 6: Detecting Stance in Tweets (https://aclanthology.org/S16-1003.pdf) from the TweetEval repo, https://github.com/cardiffnlp/tweeteval which also let me to BERTweet  (https://arxiv.org/abs/2005.10200) which scored 2nd best (but nothing scored great) on Stance Detection, but I couldn't figure out how to apply any of the outputs of that model. I am not an academic and just looking for a practical approach that uses SpaCy if possible.

UPDATE: A professor colleague of mine referred me to this article: https://aclanthology.org/2021.eacl-main.227.pdf 

And also: https://rachithaiyappa.github.io/science/Zero-Shot-for-Stance-Detection/

**Comment by metalaureate** [comment, marked as answer]
This is the solution I found: https://huggingface.co/spaces/rachith/ZeroShot_StanceDetection and outperforms the SemEval2016 benchmark without any special BERT training.

**Reply by ljvmiranda921** [reply]
Hi @metalaureate ,

For your particular problem, you can use [BERTweet](https://huggingface.co/docs/transformers/model_doc/bertweet) as a source of vectors for your task. This effectively replaces the default tok2vec component in spaCy. In theory, this should be able to better detect entities in the "tweet" domain. The [spacy-transformers](https://spacy.io/universe/project/spacy-transformers) package should be ideal for this. Within spaCy, I think using components such as [`ner`](https://spacy.io/api/entityrecognizer) and [`spacat`](https://spacy.io/api/spancategorizer) may be of help. They do a bit of structure prediction and sequence labelling, and you can access their predictions in the `Doc.ents` and `Doc.spans` attributes. 

Note that this approach only allows you to use the vectors in BERTweet for your downstream task (ner, textcat, etc.), you still have to come up with an approach on how to combine them to detect actual stances. The [paper you linked](https://aclanthology.org/S16-1003.pdf) has a nice discussion on Section 5.3 on how other teams combined different NLP tasks to do stance marking.
